policy-controller/src/main.rs

```diff
@@ -0,0 +1,148 @@
+
+    let (drain_tx, drain_rx) = drain::channel();
+
+    let client = kube::Client::try_default()
+        .await
+        .context("failed to initialize kubernetes client")?;
+
+    let (ready_tx, ready_rx) = watch::channel(false);
+    let admin = tokio::spawn(linkerd_policy_controller::admin::serve(
+        admin_addr, ready_rx,
+    ));
+
+    const DETECT_TIMEOUT: time::Duration = time::Duration::from_secs(10);
+    let (handle, index_task) = linkerd_policy_controller::k8s::index(
+        client,
+        ready_tx,
+        cluster_networks,
+        identity_domain,
+        default_allow,
+        DETECT_TIMEOUT,
+    );
+    let index_task = tokio::spawn(index_task);
+
+    let grpc = tokio::spawn(grpc(grpc_addr, handle, drain_rx));
+
+    tokio::select! {
+       _ = shutdown(drain_tx) => Ok(()),
+       res = grpc => match res {
+           Ok(res) => res.context("grpc server failed"),
+           Err(e) if e.is_cancelled() => Ok(()),
+           Err(e) => Err(e).context("grpc server panicked"),
+       },
+       res = index_task => match res {
+           Ok(e) => Err(e).context("indexer failed"),
+           Err(e) if e.is_cancelled() => Ok(()),
+           Err(e) => Err(e).context("indexer panicked"),
+       },
+       res = admin => match res {
+           Ok(res) => res.context("admin server failed"),
+           Err(e) if e.is_cancelled() => Ok(()),
+           Err(e) => Err(e).context("admin server panicked"),
+       },
+    }
+}
+
+#[derive(Debug)]
+struct IpNets(Vec<IpNet>);
+
+impl std::str::FromStr for IpNets {
+    type Err = anyhow::Error;
+    fn from_str(s: &str) -> Result<Self> {
+        s.split(',')
+            .map(|n| n.parse().map_err(Into::into))
+            .collect::<Result<Vec<IpNet>>>()
+            .map(Self)
+    }
+}
+
+#[instrument(skip(handle, drain))]
+async fn grpc(
+    addr: SocketAddr,
+    handle: linkerd_policy_controller_k8s_index::Reader,
+    drain: drain::Watch,
+) -> Result<()> {
+    let server = linkerd_policy_controller_grpc::Server::new(handle, drain.clone());
+    let (close_tx, close_rx) = tokio::sync::oneshot::channel();
+    tokio::pin! {
+        let srv = server.serve(addr, close_rx.map(|_| {}));
+    }
+    info!(%addr, "gRPC server listening");
+    tokio::select! {
+        res = (&mut srv) => res?,
+        handle = drain.signaled() => {
+            let _ = close_tx.send(());
+            handle.release_after(srv).await?
+        }
+    }
+    Ok(())
+}
+
+async fn shutdown(drain: drain::Signal) {
+    tokio::select! {
+        _ = tokio::signal::ctrl_c() => {
+            debug!("Received ctrl-c");
+        },
+        _ = sigterm() => {
+            debug!("Received SIGTERM");
+        }
+    }
+    info!("Shutting down");
+    drain.drain().await;
+}
+
+async fn sigterm() {
+    use tokio::signal::unix::{signal, SignalKind};
+    match signal(SignalKind::terminate()) {
+        Ok(mut term) => term.recv().await,
+        _ => future::pending().await,
+    };
+}
```

Comment:
should this be `cfg(unix)`'d, for people trying to e.g. run the unit tests on Windows?

Reply:
there isn't any expectation that our controllers run on windows